src/v/rpc/parse_utils.h

```diff
@@ -82,7 +83,187 @@ inline void validate_payload_and_header(const iobuf& io, const header& h) {
+            } else {
+                return ss::do_with(
+                  std::move(msg), [&out, version](const T& msg) {
+                      return serde::write_async(out, msg).then(
+                        [version] { return version; });
+                  });
+            }
+        }
+    } else {
+        /*
+         * TODO: remove this else block as soon as all types are supported by
+         * serde. this exception is present so that work on the protocol can
+         * proceed in parallel with conversion of message types to use serde.
+         *
+         * Tracked here as part of v22.2.x release requirements:
+         * https://github.com/redpanda-data/redpanda/issues/4934
+         */
+        return reflection::async_adl<T>{}.to(out, std::move(msg)).then([] {
+            return transport_version::v0;
+        });
+    }
+}
+
+/*
+ * Decode a client request at the given transport version.
+ */
+template<typename T>
+ss::future<T>
+decode_with_version(iobuf_parser& parser, transport_version version) {
+    if constexpr (serde::supported<T>) {
+        if constexpr (is_rpc_adl_exempt<T>) {
+            if (version < transport_version::v2) {
+                return ss::make_exception_future<T>(
+                  std::runtime_error(fmt::format(
+                    "Unexpected serde-only message {} at {} < v2",
+                    typeid(T).name(),
+                    version)));
+            }
+            return serde::read_async<T>(parser);
+        } else {
+            if (version < transport_version::v2) {
+                return reflection::async_adl<T>{}.from(parser);
+            } else {
+                return serde::read_async<T>(parser);
+            }
+        }
+    } else {
+        /*
+         * TODO: remove this else block as soon as all types are supported by
+         * serde. this exception is present so that work on the protocol can
+         * proceed in parallel with conversion of message types to use serde.
+         *
+         * Tracked here as part of v22.2.x release requirements:
+         * https://github.com/redpanda-data/redpanda/issues/4934
+         */
+        if (version != transport_version::v0) {
+            return ss::make_exception_future<T>(std::runtime_error(fmt::format(
+              "Unexpected adl-only message {} at {} != v0",
+              typeid(T).name(),
+              version)));
+        }
+        return reflection::async_adl<T>{}.from(parser);
+    }
+}
+
+/*
+ * type used to factor out version-specific functionality from request handling
+ * in services. this is used so that tests can specialize behavior.
+ *
+ * this is the default mixin that is used by the code generator.
+ */
+struct default_message_encoder {
+    template<typename T>
+    static ss::future<T>
+    decode(iobuf_parser& parser, transport_version version) {
+        return decode_with_version<T>(parser, version);
+    }
+
+    static transport_version version(const header& h) { return h.version; }
```

Comment:
Kinda hard to track all kinds of versions involved. I think this is a `response_version` based on `request_header`?

Reply:
it is, yes. i'll change the name. the important thing to remember is that there is a version decided upon when the request starts, and that version should remain unchanged for the lifetime of the request--through the server and back to the client as a response.

Reply:
fixed. also added some comments to the other methods in the codec type.